File: NeetCode/Stack/Maximum_Frequency/Maximum_Frequency.cpp
// Maximum_Frequency.cpp
//
// Compile locally with:
// clang++ Maximum_Frequency.cpp -std=c++17 -Wall -Wextra -O2 -o run && ./run
//
// NOTE:
// - Remove main() before submitting to LeetCode.
// - Keep only the class Solution definition.
//

#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

class FreqStack
{
private:
    unordered_map<int, int> freq;
    unordered_map<int, vector<int>> groups;
    int maxFreq;

public:
    FreqStack()
    {
        maxFreq = 0;
    }

    void push(int val)
    {
        freq[val]++;
        groups[freq[val]].push_back(val);
        maxFreq = max(maxFreq, freq[val]);
    }

    int pop()
    {
        int val = groups[maxFreq].back();
        groups[maxFreq].pop_back();
        freq[val]--;
        if (groups[maxFreq].size() == 0)
        {
            maxFreq--;
        }
        return val;
    }
};

int main()
{
    cout << "Running Maximum_Frequency.cpp..." << endl;

    FreqStack sol;

    // TODO:
    // Add local test calls here

    return 0;
}

File: NeetCode/Stack/Maximum_Frequency/maximum__frequency.py
from typing import Optional, List
import heapq
from collections import defaultdict


class FreqStack:

    def __init__(self):
        self.freq = defaultdict(int)
        self.groups = defaultdict(list)
        self.maxFreq = 0

    def push(self, val: int) -> None:
        self.freq[val] += 1
        count = self.freq[val]

        self.groups[count].append(val)
        self.maxFreq = max(self.maxFreq, count)

    def pop(self) -> int:
        val = self.groups[self.maxFreq].pop()
        self.freq[val] -= 1
        if not self.groups[self.maxFreq]:
            self.maxFreq -= 1
        return val


if __name__ == "__main__":
    sol = FreqStack()
    print("Running Solution...")


""" 
# solution works, but in O(log N) time and O(N) space
class FreqStack:

    def __init__(self):
        self.hq = []
        self.counter = defaultdict(int)
        self.order = 0

    def push(self, val: int) -> None:
        self.counter[val] += 1
        self.order += 1
        heapq.heappush(self.hq, (-self.counter[val], -self.order, val))

    def pop(self) -> int:
        count, order, val = heapq.heappop(self.hq)
        self.counter[val] -= 1
        return val

"""
